### PR TITLE
[#170129433] Add functions to get user

### PR DESCRIPTION
**Pull request by alexgpeppe**
This PR aims to add a new function to get the user information, that is the subscription list and the group list of the User identified by the provided email.

### CODE REVIEW

File: GetUser/handler.ts [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+      })
+      .chain(
+        fromPredicate(
+          userInfo => isRight(userInfo.errorOrGroups),
+          userInfoWithError =>
+            genericInternalErrorHandler(
+              context,
+              "GetUser | Invalid group contract from APIM.",
+              userInfoWithError.errorOrGroups.value as Error,
+              genericErrorMessage
```
Comment (gunzip) [inline GetUser/handler.ts]: we should return the full error message to the api caller (Invalid group contract from APIM.), same for other calls of genericInternalErrorHandler.

File: utils/conversions.ts [hunk 2]
```diff
@@ -124,3 +133,56 @@ export function userContractToApiUser(
...
+      secondary_key: subscription.secondaryKey,
+      start_date: subscription.startDate,
+      state: subscription.state,
+      state_comment: subscription.stateComment,
+      type: subscription.type
+    })
+  ).mapLeft(errors => new Error(errorsToReadableMessages(errors).join(" / ")));
+}
+
+function removeNullProperties<T>(obj: T): unknown {
```
Comment (gunzip) [inline utils/conversions.ts]: why do we need this ? can we just return the object with eventually null values ?

File: utils/middlewares/email.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import { EmailAddress } from "io-functions-commons/dist/generated/definitions/EmailAddress";
+import { RequiredParamMiddleware } from "io-functions-commons/dist/src/utils/middlewares/required_param";
+
+/**
+ * A middleware that extracts the serviceid value from the URL path parameter.
+ */
+export const EmailMiddleware = RequiredParamMiddleware("email", EmailAddress);
```
Comment (gunzip) [inline utils/middlewares/email.ts]: just use RequiredParamMiddleware("email", EmailAddress); inline in the handler, do not create a new middleware here

**Comment by alexgpeppe** [comment]
Yes, @gunzip: this PR is currently in a `draft` state because it lacks the tests, I haven't finished writing them yet.

File: utils/conversions.ts (hunk 2 above)
Comment (alexgpeppe) [reply utils/conversions.ts]: The object returned here will be used to create a response payload. The type and codec of the payload are generated from the specs file, which is written in swagger format, and this format does not support the `null` value.

**Comment by BurnedMarshal** [comment]
# [Codecov](https://codecov.io/gh/pagopa/io-functions-admin/pull/22?src=pr&el=h1) Report
> Merging [#22](https://codecov.io/gh/pagopa/io-functions-admin/pull/22?src=pr&el=desc) into [master](https://codecov.io/gh/pagopa/io-functions-admin/commit/f0268cbfb34ce1cd2b0ab769bf9cac7bbf9b77d9&el=desc) will **increase** coverage by `1.50%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/pagopa/io-functions-admin/pull/22?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master      #22      +/-   ##
==========================================
+ Coverage   85.45%   86.95%   +1.50%     
==========================================
  Files          15       17       +2     
  Lines         330      437     +107     
  Branches       26       30       +4     
==========================================
+ Hits          282      380      +98     
- Misses         47       56       +9     
  Partials        1        1              
```


| [Impacted Files](https://codecov.io/gh/pagopa/io-functions-admin/pull/22?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [GetUser/handler.ts](https://codecov.io/gh/pagopa/io-functions-admin/pull/22/diff?src=pr&el=tree#diff-R2V0VXNlci9oYW5kbGVyLnRz) | `92.40% <0.00%> (ø)` | |
| [utils/errorHandler.ts](https://codecov.io/gh/pagopa/io-functions-admin/pull/22/diff?src=pr&el=tree#diff-dXRpbHMvZXJyb3JIYW5kbGVyLnRz) | `84.61% <0.00%> (ø)` | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/pagopa/io-functions-admin/pull/22?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/pagopa/io-functions-admin/pull/22?src=pr&el=footer). Last update [f0268cb...f58277c](https://codecov.io/gh/pagopa/io-functions-admin/pull/22?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

File: GetUser/__tests__/handler.test.ts [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+const spyOnGetApiClient = jest.spyOn(ApimUtils, "getApiClient");
+spyOnGetApiClient.mockImplementation(() =>
+  fromEither(right(new mockApiManagementClient()))
+);
+
+const mockLog = jest.fn();
+const mockedContext = { log: { error: mockLog } };
+
+describe("GetUser", () => {
+  it("should return an internal error response if the API management client can not be gotten", async () => {
```
Comment (gunzip) [inline GetUser/__tests__/handler.test.ts]:
```suggestion
  it("should return an internal error response if the API management client can not be got", async () => {
```

File: GetUser/__tests__/handler.test.ts [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+      undefined as any,
+      undefined as any,
+      undefined as any
+    );
+
+    expect(response.kind).toEqual("IResponseErrorInternal");
+  });
+});
+
+describe("GetUser", () => {
```
Comment (gunzip) [inline GetUser/__tests__/handler.test.ts]: put everything under just one describe call as the test suite name is the same

File: openapi/index.yaml [hunk 6]
```diff
@@ -394,6 +419,96 @@ definitions:
...
+      group_contract_type:
+        $ref: "#/definitions/GroupType"
+      id:
+        type: string
+      name:
+        type: string
+      type:
+        type: string
+    required:
+      - display_name
```
Comment (gunzip) [inline openapi/index.yaml]:
```suggestion
      display_name:
        type: string
      id:
        type: string
      name:
        type: string
    required:
      - display_name
```
we don't need all the group fields for this API, just return a subset
(you may need to tweak groupContractToApiGroup)

File: openapi/index.yaml [hunk 7]
```diff
@@ -394,6 +419,96 @@ definitions:
...
+  GroupType:
+    type: string
+    x-extensible-enum:
+      - custom
+      - system
+      - external
```
Comment (gunzip) [inline openapi/index.yaml]: we can safely remove this

File: openapi/index.yaml [hunk 8]
```diff
@@ -394,6 +419,96 @@ definitions:
...
+        type: string
+      start_date:
+        type: string
+        format: "date-time"
+      state:
+        $ref: "#/definitions/SubscriptionState"
+      state_comment:
+        type: string
+      type:
+        type: string
```
Comment (gunzip) [inline openapi/index.yaml]:
```suggestion
      allow_tracing:
        type: boolean
      created_date:
        type: string
        format: "date-time"
      display_name:
        type: string
      end_date:
        type: string
        format: "date-time"
      expiration_date:
        type: string
        format: "date-time"
      id:
        type: string
      name:
        type: string
      notification_date:
        type: string
      owner_id:
        type: string
      scope:
        type: string
      start_date:
        type: string
        format: "date-time"
      state:
        $ref: "#/definitions/SubscriptionState"
      state_comment:
        type: string
      type:
        type: string
```

Same suggestion as for Groups. We need to hide the keys here, we will retrieve them by a subsequent call to
GetSubscriptionKeys(subscriptionId).

File: openapi/index.yaml [hunk 9]
```diff
@@ -394,6 +419,76 @@ definitions:
...
+    required:
+      - primary_key
```
Comment (gunzip) [inline openapi/index.yaml]:
```suggestion
```

File: openapi/index.yaml [hunk 10]
```diff
@@ -394,6 +419,76 @@ definitions:
...
+      - scope
+      - secondary_key
```
Comment (gunzip) [inline openapi/index.yaml]:
```suggestion
```

**Comment by BurnedMarshal** [comment]
# [Codecov](https://codecov.io/gh/pagopa/io-functions-admin/pull/22?src=pr&el=h1) Report
> Merging [#22](https://codecov.io/gh/pagopa/io-functions-admin/pull/22?src=pr&el=desc) into [master](https://codecov.io/gh/pagopa/io-functions-admin/commit/1851bc550ea24b980cf0055235e3d2ecca50197f&el=desc) will **increase** coverage by `1.50%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/pagopa/io-functions-admin/pull/22?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master      #22      +/-   ##
==========================================
+ Coverage   85.45%   86.95%   +1.50%     
==========================================
  Files          15       17       +2     
  Lines         330      437     +107     
  Branches       26       30       +4     
==========================================
+ Hits          282      380      +98     
- Misses         47       56       +9     
  Partials        1        1              
```


| [Impacted Files](https://codecov.io/gh/pagopa/io-functions-admin/pull/22?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [GetUser/handler.ts](https://codecov.io/gh/pagopa/io-functions-admin/pull/22/diff?src=pr&el=tree#diff-R2V0VXNlci9oYW5kbGVyLnRz) | `92.40% <0.00%> (ø)` | |
| [utils/errorHandler.ts](https://codecov.io/gh/pagopa/io-functions-admin/pull/22/diff?src=pr&el=tree#diff-dXRpbHMvZXJyb3JIYW5kbGVyLnRz) | `84.61% <0.00%> (ø)` | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/pagopa/io-functions-admin/pull/22?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/pagopa/io-functions-admin/pull/22?src=pr&el=footer). Last update [1851bc5...feb5c16](https://codecov.io/gh/pagopa/io-functions-admin/pull/22?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

**Pull request merged**